### PR TITLE
Allow caching null objects.

### LINKED ISSUE #939: cache.put("key",null) fails [DATACOUCH-628]
**[achoice](https://jira.spring.io/secure/ViewProfile.jspa?name=JIRAUSER50333)** opened **[DATACOUCH-628](https://jira.spring.io/browse/DATACOUCH-628?redirect=false)** and commented

Caching of null values fail with exception thrown by Couchbase client.

CouchbaseCacheConfiguration.defaultCacheConfig() seems that null valus should be cachable. https://github.com/spring-projects/spring-data-couchbase/blob/master/src/main/java/org/springframework/data/couchbase/cache/CouchbaseCacheConfiguration.java#L54

 

```java
@Test
public void cacheNull() {
    final String CACHE_NAME = "CACHE_TEST";
    
    final CouchbaseCacheConfiguration defaultCacheConfig = CouchbaseCacheConfiguration.defaultCacheConfig();
    assertTrue(defaultCacheConfig.getAllowCacheNullValues()); // expect null values to be cachable now

    final CouchbaseCacheManager cacheManager = CouchbaseCacheManager
            .builder(couchbaseClientFactory)
            .cacheDefaults(defaultCacheConfig)
            .initialCacheNames(Set.of(CACHE_NAME))
            .build();
    cacheManager.afterPropertiesSet();

    // TEST cache null
    final Cache cache = cacheManager.getCache(CACHE_NAME);
    cache.put("KEY", null); // exception thrown by couchbase client
}
```

 

---

**Affects:** 4.0.4 (Neumann SR4)

### PR DESCRIPTION
**Pull request by mikereiche**
Use toStoreValue(value) to substitute null objects with NullValue.INSTANCE when putting.  fromStoreValue(value) does the opposite when getting.

Closes #939.

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATACOUCH).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

File: src/test/java/org/springframework/data/couchbase/cache/CacheUser.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * limitations under the License.
+ */
+package org.springframework.data.couchbase.cache;
+
+import java.io.Serializable;
+
+/**
+ * This is a standalone class (vs. inner) to allow Serialization of all fields to work.
+ * If it was an inner class of CouchbaseCacheIntegrationTests, then it would have a
+ * this$0 field = CouchbaseCacheIntegrationTests and would not serialize.
```
Comment (dnault) [inline src/test/java/org/springframework/data/couchbase/cache/CacheUser.java:23]: If CacheUser were a static nested class, would it it still need to be in a separate file?

https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html

**Review by dnault — APPROVED** [review]
LGTM, with some nits.

File: src/test/java/org/springframework/data/couchbase/cache/CouchbaseCacheIntegrationTests.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+import com.couchbase.client.java.query.QueryOptions;
+
+/**
+ * CouchbaseCache tests Theses tests rely on a cb server running.
+ *
+ * @author Michael Reiche
+ */
+@IgnoreWhen(clusterTypes = ClusterType.MOCKED)
+class CouchbaseCacheIntegrationTests extends JavaIntegrationTests {
```
Comment (dnault) [inline src/test/java/org/springframework/data/couchbase/cache/CouchbaseCacheIntegrationTests.java:40]: These tests are good. Are they related to the "Allow caching null objects" change?

Should we have a test that caches and retrieves a null value?

**Review by dnault — APPROVED** [review]
LGTM, with some nits.

File: src/test/java/org/springframework/data/couchbase/cache/CouchbaseCacheIntegrationTests.java (hunk 2 above)
Comment (mikereiche) [reply src/test/java/org/springframework/data/couchbase/cache/CouchbaseCacheIntegrationTests.java:40]: cacheHitMiss() tests a cache hit on a stored null.
It also tests a cache-miss - which is the same for a null as a non-null.
The other tests have cache-hits on non-null (the normal case).
